homeassistant/components/enocean/__init__.py

```diff
@@ -1,93 +1,74 @@
 )
 
-SIGNAL_RECEIVE_MESSAGE = "enocean.receive_message"
-SIGNAL_SEND_MESSAGE = "enocean.send_message"
-
 
-def setup(hass, config):
+async def async_setup(hass, config):
     """Set up the EnOcean component."""
-    serial_dev = config[DOMAIN].get(CONF_DEVICE)
-    dongle = EnOceanDongle(hass, serial_dev)
-    hass.data[DATA_ENOCEAN] = dongle
-
-    return True
+    # support for text-based configuration (legacy)
+    if DOMAIN not in config:
+        return True
 
+    hass.data.setdefault(DATA_ENOCEAN, {})
 
-class EnOceanDongle:
-    """Representation of an EnOcean dongle."""
+    if hass.config_entries.async_entries(DOMAIN):
+        # We can only have one dongle. If there is already one in the config,
+        # there is no need to import the yaml based config.
+        return True
 
-    def __init__(self, hass, ser):
-        """Initialize the EnOcean dongle."""
-
-        self.__communicator = SerialCommunicator(port=ser, callback=self.callback)
-        self.__communicator.start()
-        self.hass = hass
-        self.hass.helpers.dispatcher.dispatcher_connect(
-            SIGNAL_SEND_MESSAGE, self._send_message_callback
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config[DOMAIN]
         )
+    )
 
-    def _send_message_callback(self, command):
-        """Send a command through the EnOcean dongle."""
-        self.__communicator.send(command)
-
-    def callback(self, packet):
-        """Handle EnOcean device's callback.
-
-        This is the callback function called by python-enocan whenever there
-        is an incoming packet.
-        """
-
-        if isinstance(packet, RadioPacket):
-            _LOGGER.debug("Received radio packet: %s", packet)
-            self.hass.helpers.dispatcher.dispatcher_send(SIGNAL_RECEIVE_MESSAGE, packet)
-
+    return True
 
-class EnOceanDevice(Entity):
-    """Parent class for all devices associated with the EnOcean component."""
 
-    def __init__(self, dev_id, dev_name="EnOcean device"):
-        """Initialize the device."""
-        self.dev_id = dev_id
-        self.dev_name = dev_name
+async def async_setup_entry(
+    hass: core.HomeAssistant, config_entry: config_entries.ConfigEntry
+):
+    """Set up an EnOcean dongle for the given entry."""
+    enocean_data = hass.data.setdefault(DATA_ENOCEAN, {})
+    usb_dongle = EnOceanDongle(hass, config_entry.data[CONF_DEVICE])
+    await usb_dongle.async_setup()
+    enocean_data[ENOCEAN_DONGLE] = usb_dongle
 
-    async def async_added_to_hass(self):
-        """Register callbacks."""
-        self.async_on_remove(
-            self.hass.helpers.dispatcher.async_dispatcher_connect(
-                SIGNAL_RECEIVE_MESSAGE, self._message_received_callback
-            )
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
```

Comment:
I think we can remove the forwarding since we will require config yaml setup for the platforms.

If we need the config_entry.data in the platforms we can store it in hass.data.

Reply:
Forwarding removed